sbot/serial_wrapper.py

```diff
@@ -0,0 +1,238 @@
+            port,
+            baudrate=baud,
+            timeout=timeout,
+            do_not_open=True,
+        )
+
+    def start(self) -> None:
+        """
+        Helper method to open the serial port.
+
+        This is not usually needed as the port will be opened on the first message.
+        """
+        self._connect()
+
+    def stop(self) -> None:
+        """
+        Helper method to close the serial port.
+
+        This is not usually needed as the port will be closed on garbage collection.
+        """
+        self._disconnect()
+
+    @retry(times=3, exceptions=BoardDisconnectionError)
+    def query(self, data: str) -> str:
+        """
+        Send a command to the board and return the response.
+
+        This method will automatically reconnect to the board and retry the command
+        up to 3 times on serial errors.
+
+        :param data: The data to write to the board.
+        :raises BoardDisconnectionError: If the serial connection fails during the transaction,
+            including failing to respond to the command.
+        :return: The response from the board with the trailing newline removed.
+        """
+        with self._lock:
+            if not self.serial.is_open:
+                if not self._connect():
+                    # If the serial port cannot be opened raise an error,
+                    # this will be caught by the retry decorator
+                    raise BoardDisconnectionError((
+                        f'Connection to board {self.identity.board_type}:'
+                        f'{self.identity.asset_tag} could not be established',
+                    ))
+
+            try:
+                logger.log(TRACE, f'Serial write - {data!r}')
+                cmd = data + '\n'
+                self.serial.write(cmd.encode())
+
+                response = self.serial.readline()
+                # Drop all unicode characters that cannot be decoded
+                response_str = response.decode(errors='ignore').rstrip('\n')
+                logger.log(
+                    TRACE, f'Serial read  - {response_str!r}')
+
+                if b'\n' not in response:
+                    # If readline times out no error is raised, it returns an incomplete string
+                    logger.warning((
+                        f'Connection to board {self.identity.board_type}:'
+                        f'{self.identity.asset_tag} timed out waiting for response'
+                    ))
+                    raise serial.SerialException('Timeout on readline')
+            except serial.SerialException:
+                # Serial connection failed, close the port and raise an error
+                self._disconnect()
+                raise BoardDisconnectionError((
+                    f'Board {self.identity.board_type}:{self.identity.asset_tag} '
+                    'disconnected during transaction'
+                ))
+
+            if 'NACK' in response_str:
```

Comment:
Nit: Just in case the error string contains "NACK"
```suggestion
            if response_str.startswith('NACK'):
```

Reply:
cb67f8a99d1d38d9530474075709322e67e17446